### DISCUSSION TITLE
Questions: swa start

### DISCUSSION THREAD

**Discussion by anthonychu** [Azure/static-web-apps-cli#383] [Q&A]
Have a few questions about parts of the `start` command that are unclear...

* What's the "context" and how does that differ from appLocation or appArtifactLocation?
    - If I have a folder named `static` and I want to serve it, it seems like the only way to make this work is to set the context: `swa start ./static`. `swa start --app-location ./static` (or using `--app`) doesn't seem to work.
* How does it select the workflow file to use, especially if there's more than one?
* What does it use from the workflow?
* How does `--build` work?
* How does it determine which config file to use? #125

**Comment by manekinekko** [comment]
Great questions! Let's talk about this during our next call and then I'll write down the answers in the wiki.

**Comment by manekinekko** [comment]
> What's the "context" and how does that differ from appLocation or appArtifactLocation?

`context` determines how the emulator will serve the static content:
- if `context` is a folder, then we will start the emulator and serve static content from that folder
- if `context` is an HTTP address, then we will start the emulator and proxy req to the dev server

In both cases, we reconfigure `appLocation` and `appArtifactLocation` to reflect that:
- `appArtifactLocation = context`
- `appLocation = cwd()` (currently, we don't use `appLocation` if `context` is a dev server)

> If I have a folder named static and I want to serve it, it seems like the only way to make this work is to set the context: swa start ./static. swa start --app-location ./static (or using --app) doesn't seem to work.

`swa start --app-location ./static` is the old way. Technically speaking this:
```
swa start --app-location ./static
```

translates to this:

```
swa start <context> --app-location ./static
```

When not provided, `context` maps to the `pwd`:

```
swa start ./ --app-location ./static
```

My understanding of how [commander](https://www.npmjs.com/package/commander) works, is that `context` (a command) will take precedence over `--app-location` (an options). But I am sure with a little of checks we can change that.

> How does it select the workflow file to use, especially if there's more than one?

The scans scans for `azure-static-web-apps-**.yml` files under `cwd()/.github/workflows/`, and picks the 1st one (non-deterministic!).

> What does it use from the workflow?

```
app_build_command
api_build_command
app_location
app_artifact_location
api_location
```

These values are overridden by user's config (CLI flags).

> How does --build work?

We check for runtime type. We currently support Node.js and .NET (experimental):
- For Node.js, 
  - For app, we invoke `npm install && <app_build_command>`
  - For api we invoke `npm install && <api_build_command>`
- For .NET, 
  - We invoke `dotnet build`

**Important: We don't build if `context` is a dev server.**

> How does it determine which config file to use

We scan recursively the `appLocation` for `staticwebapp.config.json` (or `routes.json`):
- If we find both `staticwebapp.config.json` and `routes.json`, we pick `staticwebapp.config.json`.
- If we find only `routes.json`, verify it's a valid SWA config file (some JS framework use same filename for routing) then pick it.

**Comment by Divyesh85** [comment]
Hi @anthonychu , i was trying to implement Static Web apps and add API to it as described in this document - https://docs.microsoft.com/en-us/azure/static-web-apps/add-api?tabs=vanilla-javascript but seems i am having problem running this command on my bash terminal. "swa start src --api-location api"
I get below error when i try to run it, i am not sure if i missed anything to install or configure. I have followed all the steps mentioned in this document already. 
![image](https://user-images.githubusercontent.com/17871562/147467083-24254d0a-fcd5-4243-a635-c572be74a429.png)
I am using Windows machine , and running the code from the root directory.

**Comment by Reshmi-Sriram** [comment]
> 

Hi @Divyesh85 ,

Sorry that you faced this issue. Could you please confirm if you followed the Installation steps mentioned [above](https://docs.microsoft.com/en-us/azure/static-web-apps/add-api?tabs=vanilla-javascript#install-command-line-tools)? One usually comes across this error when the SWA CLI is not installed, hence the bash is not able to recognise the command `swa`

Also, it would be great if you could help us repro the error by sharing your project in a public repository if possible.
Hope this helps!

**Comment by Divyesh85** [comment]
@Reshmi-Sriram sorry for late reply due to holidays, i was able to resolve the issue, seems by default when i was trying to install azure functions and azure static webapps cli on VS code terminal it was not installing using the administrator mode hence was not able to install properly. I tried it in separate command prompt as an admin and it went well. Thanks for the prompt reply though.
